### ISSUE TITLE
allow applications to use SO_REUSEPORT

### ISSUE THREAD

**Issue by marten-seemann** [lucas-clemente/quic-go#2597]
SO_REUSEPORT allows us to create multiple listeners for a single address / port.

I created a simple script to test what's going on: https://gist.github.com/marten-seemann/42199c0b8462c1a5748c6b1c1aca22c4

It looks like the kernel will deterministically route packets coming from a certain remote address to a single listener. That means that this won't show up as additional reordering (see https://github.com/lucas-clemente/quic-go/issues/2586#issuecomment-641171225).

This means that reuseport could help us improve performance when handling *multiple* connections. It won't help with single-connection performance, since in that case only a single socket is used. Further benchmarking is needed to evaluate how big the performance gain is for quic-go in specific, but https://blog.cloudflare.com/how-to-receive-a-million-packets/ suggest that this could be worth doing.

**Comment by astrolox** [comment]
FYI: This is needed in order to use the autocert module to automatically provision LetsEncrypt certificates.
https://pkg.go.dev/golang.org/x/crypto/acme/autocert

**Comment by marten-seemann** [comment]
We need to be very careful here. This might break connection migration, as a client that is migrating would send from a different remote address. As far as I can see, while the kernel makes sure that packets sent from the same remote address end up at the same connection, it doesn't have any concept of QUIC connections, so there's no guarantee that packets will be received by the original connection.
If they end up at another connection, they'll get routed to a different `packetHandlerMap`, and trigger a Stateless Reset there.

**Comment by rip-create-your-account** [comment]
For server scalability this would be important. Artificial benchmark with a high number of clients sending in parallel has horrible results as the packets sent by clients end up fighting for the single kernelspace lock of the socket.

For `SO_REUSEPORT` there exists `SO_ATTACH_REUSEPORT_EBPF` (Linux 4.5) which could then be used to consistently direct received packets to the same listener socket based on the connection id in the packet (instead of the flow hash that Linux would otherwise use). I think this would solve the issue with clients doing connection migration.

Also in userspace the server transmission path has a parallelism bottleneck in the form of [fdMutex](https://github.com/golang/go/blob/44361140c02556a0a71bc52299149bb8de26024b/src/internal/poll/fd_unix.go#L331) that serializes the writes to the socket. Packets sent to two different clients are not sent in parallel. This could be fixed by Dup()-ing the PacketConn via `(*net.UDPConn).File()` for every client but with the downside that the process then may reach the maximum number of file descriptors open. Then the kernel locks in transmit path would likely again kill the parallelism so this doesn't really help. This may be worth a separate issue.

**Comment by rip-create-your-account** [comment]
I implemented a very crude and hacky version of `SO_REUSEPORT` with `SO_ATTACH_REUSEPORT_EBPF`. Server scalability improved significantly when large number of clients are generating traffic in parallel. For the case of only one client there was no noticeable performance degradation. 

The benchmark is artificial worst case. For only 8 clients `perf` reports ~3% of overhead from fighting for the kernelspace lock without `SO_REUSEPORT`. For 16 clients it reports ~8% and for 100 clients ~40% of overhead. The CPU I tested on has only 8 cores. I assume that having more cores would increase the overhead significantly. With `SO_REUSEPORT` and 8 server sockets the reported overhead was ~0.2% for 100 clients. This was realized as 2.5x throughput.

I modified quic-go to allow the user to call `quic.Listen` for the same address multiple times. Each time the user would then pass in a different `socket` but always bound to the same address with `SO_REUSEPORT` and also give "index" for use of the `EBPF` program.

I wrote the `EBPF` program to use the first byte of the destination CID in the packet to route it to the correct socket. `Listener` and the `Sessions` it has created would then generate CIDs with the first byte unique to that listener. This unique byte is the "index" passed to `quic.Listen`. I tested with 8 `quic.Listeners` so to keep things simple the "index" was just a number from 0 to 7. Then the `EBPF` program could just use a bitmask to find the correct socket:
```c
socket_index = first_byte_of_dcid & 0b111; // assume 8 sockets
bpf_sk_select_reuseport(reuse_md, &sockets_map, &socket_index, 0 /*flags*/);
```
The `EBPF` program understands both long and short header packets. Because clients generate random destination CID for the initial packet it would mean that a `Listener` matching the random first byte would receive that packet. From that point onwards that `Listener` is able to generate CIDs so that the client would consistently have it's packets routed to it. For my benchmark I modified the clients to generate the initial destination CIDs so that the load would be equally distributed.

One of the many issues that I didn't solve here is how to route a stateless reset sent by the _client_. Such packets don't have a real destination CID so with my hacky implementation the wrong `packetHandlerMap` would end up receiving it.

The downside of `SO_ATTACH_REUSEPORT_EBPF` is that loading the `EBPF` program requires `CAP_BPF`. And it seems to exists only on Linux.

To me my test has proved that this is a worthwhile feature as it fixes the kernelspace bottleneck.

Thank you for coming to my ted talk

**Comment by marten-seemann** [comment]
That's super interesting! Thank you for digging into this!

> The benchmark is artificial worst case. For only 8 clients perf reports ~3% of overhead from fighting for the kernelspace lock without SO_REUSEPORT. For 16 clients it reports ~8% and for 100 clients ~40% of overhead. 

Can you share how you do these tests? How do you determine the kernel lock overhead?

> The downside of SO_ATTACH_REUSEPORT_EBPF is that loading the EBPF program requires CAP_BPF. And it seems to exists only on Linux.

That this only works on Linux is expected. quic-go is not guaranteed to perform well on any other system anyway ;)
Does CAP_BPF mean that you'd have to run the binary as root?

> One of the many issues that I didn't solve here is how to route a stateless reset sent by the client. Such packets don't have a real destination CID so with my hacky implementation the wrong packetHandlerMap would end up receiving it.

Do we need a separate `packetHandlerMap` per socket? How about using the same one, with multiple `listen` go routines? Would that create too much mutex contention?

**Comment by rip-create-your-account** [comment]
> Can you share how you do these tests?

I have a heavily modified version of the benchmark in `benchmarks/benchmark_test.go`. It's not really respective of the real world as the client and server both run on the same machine but I think it's good enough to test this. The test would spin up multiple clients and wait until they have all finished the handshake. Then the clients would open a stream and send. The benchmark does a lot of unrelated stuff so I don't want to share the code as is.

> How do you determine the kernel lock overhead?

I run the benchmark with `perf record` and then afterwards view the results with `perf report`. The report shows in the UDP receive path the overhead for `native_queued_spin_lock_slowpath`. I simply used the reported overhead % for it.

> Does CAP_BPF mean that you'd have to run the binary as root?

Yes, or add the CAP_BPF to the binary with `sudo setcap cap_bpf+ep quic_binary`.

> Do we need a separate packetHandlerMap per socket? How about using the same one, with multiple listen go routines? Would that create too much mutex contention?

I think that would then lead to mutex contention when accessing the `handlers`. Maybe the stateless reset map could be separated from `packetHandlerMap` and then have the many `packetHandlerMaps` share it. I think here it's also important what information the `EBPF` program has to route the packets to the correct socket. In my test it knows that the first byte of the destination CID tells the socket. I think one other way would be to have a `EBPF map` in the program which maps a CID to a socket. But this would mean map access in the `EBPF` program which then may be its own bottleneck. Also it would require quic-go to make a syscall to add/remove the generated CIDs to/from that map. This would make handling the initial packets cumbersome as there's no clear time to remove the entries. Also, I'm no expert on `EBPF` or its internals.

The proper way of how the `EBPF` program and quic-go share state I have no conclusive idea of. There exists [QUIC-LB](https://datatracker.ietf.org/doc/html/draft-ietf-quic-load-balancers-06) which I think has useful ideas how this could be done properly. Dedicating the first byte to be the index I think is not the proper way.

**Comment by rip-create-your-account** [comment]
> Maybe the stateless reset map could be separated from `packetHandlerMap` and then have the many `packetHandlerMaps` share it

Currently `packetHandlerMap` [checks](https://github.com/lucas-clemente/quic-go/blob/622ca23d4eb4a3b0b87452742653cdf094fd9ecf/packet_handler_map.go#L318) every packet in case it's a Stateless Reset before handing it off to a `packetHandler`. A separate shared map of Stateless Reset handlers would just become the new bottleneck. The obvious fix would be to check if a packet is a Stateless Reset _after_ trying to find a matching `packetHandler`. But because Stateless Resets don't have a real destination CID they could accidentally match a `packetHandler` of unrelated Session. Also the `ConnectionIDLength` could be configured to 0 which would mean that this would happen for every Stateless Reset. To fix that the `packetHandler` could check if the packet is a Stateless Reset for some other `Session` if it was not successfully processed.

[quic-transport-34](https://tools.ietf.org/html/draft-ietf-quic-transport-34#section-10.3.1) says the following about detecting a Stateless Reset:

> This comparison can be performed for every inbound datagram.
   Endpoints MAY skip this check if any packet from a datagram is
   successfully processed.  However, the comparison MUST be performed
   when the first packet in an incoming datagram either cannot be
   associated with a connection, or cannot be decrypted.

**Comment by marten-seemann** [comment]
> The proper way of how the EBPF program and quic-go share state I have no conclusive idea of. There exists QUIC-LB which I think has useful ideas how this could be done properly. Dedicating the first byte to be the index I think is not the proper way.

QUIC-LB standardizes a coordination protocol between a load balancer that forwards packets to multiple backend servers. It also makes use of the connection ID to encode information about the server instance.
Given that using `SO_ATTACH_REUSEPORT_EBPF` is a local operation, we don't need any standardized coordination protocol. Using a part of the connection ID to encode such information seems like a totally reasonable thing to do.

> Currently packetHandlerMap checks every packet in case it's a Stateless Reset before handing it off to a packetHandler. A separate shared map of Stateless Reset handlers would just become the new bottleneck. The obvious fix would be to check if a packet is a Stateless Reset after trying to find a matching packetHandler. But because Stateless Resets don't have a real destination CID they could accidentally match a packetHandler of unrelated Session. 

I think we can live with that, and we can safely ignore the MUST in the spec ;) The reason is the following: The chance that a stateless reset matches an existing session is quite small, for 4-byte connection IDs it's `N` in 4 billion, where `N` is the number of connection IDs currently in use. In practice, this value will be a lot less than 100,000. The chance is thus much smaller than the typical packet loss rates on the public internet.
Therefore, it should be sufficient to only check for a stateless reset if a packet can't be associated with any existing session. Under normal operation, this check should be pretty rare, so we shouldn't create a lot of mutex contention if all the `packetHandlerMap`s share a single stateless reset table.

**Comment by rip-create-your-account** [comment]
QUIC-LB provides quite a bit of insight as it solves a lot of the same problems. It does bring up the issue of linkability. What I have tested with has been equal to the "Plaintext CID Algorithm" so that issue is there. I'm not a security expert or even a novice so I have no real understanding of how relevant the security issues mentioned in QUIC-LB are.

> 4-byte connection ID

If 1 byte is used for `SO_ATTACH_REUSEPORT_EBPF` the chance increases somewhat. I would assume that it's still enough to keep the number of unlucky clients low. Also I just now noticed that `(*Config).ConnectionIDLength` is not validated so it could be configured to 1, 2 or 3 which the API says not to do.

For my test I technically used the lower 3 bits (`first_byte_of_dcid & 0b111`) to encode the information, not the entire byte. I wonder how many bits actually is needed. It seems that socket per CPU core (or NIC rx-queue?) would be the best idea so that in mind 3 bits might not be enough for servers. If the number of sockets is known before loading the `EBPF` program then [the constant bitmask can be changed to the optimal value](https://pkg.go.dev/github.com/cilium/ebpf#CollectionSpec.RewriteConstants) and have it compiled as a constant by the BPF JIT.

> Therefore, it should be sufficient to only check for a stateless reset if a packet can't be associated with any existing session.

With `(*Config).ConnectionIDLength` of 0 the stateless reset would always end up being associated with the existing session.

Do you know what the API for quic-go could look like? I think that `SO_REUSEPORT` has such a great impact on how well multiple cores are utilized that it should be the default and happen automatically.

I looked into things a bit more and concluded that `UDP_GRO` works just fine with `SO_ATTACH_REUSEPORT_EBPF`. That's nice.

**Comment by rip-create-your-account** [comment]
Digging deeper to the `perf report` makes me wonder if there's more to do to scale well. With `SO_REUSEPORT` and 100 clients roughly 14% of overhead comes from `ep_poll_callback` in the kernel receive path. Is it called for every packet that is received? `UDP_GRO` I think would help in that case. But here it seems like that there's lock contention as `ep_poll_callback` locks the epoll instance, if I have understood things correctly the go runtime has only one epoll instance. Although here the overhead might be inflated by my test which has the clients and server run in the same process. When there's only 1 client then that overhead falls down to 1%. I think the fix for epoll would be more epoll instances. Perhaps `UDP_GRO` would mask the issue.
